Kernel/Interrupts/APIC.h

```diff
@@ -27,16 +27,72 @@
 #pragma once
 
 #include <AK/Types.h>
+#include <Kernel/VM/MemoryManager.h>
 
 namespace Kernel {
 
-namespace APIC {
+class APIC {
+public:
+    static APIC& the();
+    static void initialize();
+    static bool initialized();
 
-void enable_bsp();
-void eoi();
-bool init();
-void enable(u32 cpu);
-u8 spurious_interrupt_vector();
-}
+    bool init_bsp();
+    void enable_bsp();
+    void eoi();
+    void enable(u32 cpu);
+    static u8 spurious_interrupt_vector();
+
+private:
+    class ICRReg {
+        u32 m_reg { 0 };
+    
+    public:
+        enum DeliveryMode {
+            Fixed = 0x0,
+            LowPriority = 0x1,
+            SMI = 0x2,
+            NMI = 0x4,
+            INIT = 0x5,
+            StartUp = 0x6,
+        };
+        enum DestinationMode {
+            Physical = 0x0,
+            Logical = 0x0,
+        };
+        enum Level {
+            DeAssert = 0x0,
+            Assert = 0x1
+        };
+        enum class TriggerMode {
+            Edge = 0x0,
+            Level = 0x1,
+        };
+        enum DestinationShorthand {
+            NoShorthand = 0x0,
+            Self = 0x1,
+            AllIncludingSelf = 0x2,
+            AllExcludingSelf = 0x3,
+        };
+    
+        ICRReg(u8 vector, DeliveryMode delivery_mode, DestinationMode destination_mode, Level level, TriggerMode trigger_mode, DestinationShorthand destination)
+            : m_reg(vector | (delivery_mode << 8) | (destination_mode << 11) | (level << 14) | (static_cast<u32>(trigger_mode) << 15) | (destination << 18))
+        {
+        }
+    
+        u32 low() const { return m_reg; }
+        u32 high() const { return 0; }
+    };
+
+    OwnPtr<Region> m_apic_base;
+    AK::Vector<OwnPtr<Region>> m_apic_ap_stacks;
```

Comment:
```suggestion
    NonnullOwnPtrVector<Region> m_apic_ap_stacks;
```

Reply:
Hm I can't figure out how to get it to compile with `NonnullOwnPtrVector`. The append method doesn't like to take a `OwnPtr<Region>` and I haven't been able to figure out how to solve it. Any hints? I looked around and at the implementations but haven't found any clues yet.

Reply:
If you have an `OwnPtr` that you trust to be non-null, you can convert it into a `NonnullOwnPtr` with `OwnPtr::release_nonnull()`. :)